src/client/dfuse/dfuse_core.c

```diff
@@ -671,204 +756,178 @@ dfuse_set_default_cont_cache_values(struct dfuse_cont *dfc)
-		} else if (rc != -DER_SUCCESS) {
-			DFUSE_TRA_ERROR(dfc, "daos_cont_open() failed: " DF_RC, DP_RC(rc));
+		if (rc != -DER_SUCCESS) {
+			if (rc == -DER_NONEXIST || rc == -DER_NO_PERM)
+				DHL_INFO(dfc, rc, "daos_cont_open() failed");
+			else
+				DHL_ERROR(dfc, rc, "daos_cont_open() failed");
 			D_GOTO(err_free, rc = daos_der2errno(rc));
 		}
+
 		rc = dfs_mount(dfp->dfp_poh, dfc->dfs_coh, dfs_flags, &dfc->dfs_ns);
 		if (rc) {
 			DHS_ERROR(dfc, rc, "dfs mount() failed");
 			D_GOTO(err_close, rc);
 		}
 
+		uuid_copy(dfc->dfc_uuid, c_info.ci_uuid);
+
 		if (dfuse_info->di_caching) {
 			rc = dfuse_cont_get_cache(dfc);
 			if (rc == ENODATA) {
-				/* If there are no container attributes then use defaults */
 				DFUSE_TRA_INFO(dfc, "Using default caching values");
 				dfuse_set_default_cont_cache_values(dfc);
-				rc = 0;
 			} else if (rc != 0) {
 				D_GOTO(err_umount, rc);
 			}
-
-		} else {
-			DFUSE_TRA_INFO(dfc, "Caching disabled");
 		}
 
 		rc = ival_add_cont_buckets(dfc);
 		if (rc != 0)
 			goto err_umount;
 
-	} else {
-		/* This is either a container where a label is set on the
-		 * command line, or one created through mkdir, in either case
-		 * the container will be mounted, and caching etc already
-		 * setup.
+		/* Check if this container has been accessed in the past and if so then reuse the
+		 * inode number.
 		 */
-		dfc->dfs_ops = &dfuse_dfs_ops;
+		D_SPIN_LOCK(&dfuse_info->di_lock);
+		d_list_for_each_entry(dfcc, &dfp->dfp_historic, dfcc_entry) {
+			if (dfcc->dfcc_ino == 0)
+				continue;
+			if (uuid_compare(dfcc->dfcc_uuid, dfc->dfc_uuid) != 0)
+				continue;
+			dfc->dfs_ino = dfcc->dfcc_ino;
+			break;
+		}
+		if (dfc->dfs_ino == 0) {
+			struct dfuse_pool *dfpp;
+
+			DFUSE_TRA_INFO(dfc, "Looking for inode");
+
+			d_list_for_each_entry(dfpp, &dfuse_info->di_pool_historic, dfp_entry) {
+				DFUSE_TRA_INFO(dfc, "Looking for inode " DF_UUID,
+					       DP_UUID(dfpp->dfp_uuid));
+
+				if (uuid_compare(dfpp->dfp_uuid, dfp->dfp_uuid) != 0)
+					continue;
+
+				d_list_for_each_entry(dfcc, &dfpp->dfp_historic, dfcc_entry) {
+					DFUSE_TRA_INFO(
+					    dfc, "Looking for inode " DF_UUID " " DF_UUID,
+					    DP_UUID(dfpp->dfp_uuid), DP_UUID(dfcc->dfcc_uuid));
+					if (dfcc->dfcc_ino == 0)
+						continue;
+					if (uuid_compare(dfcc->dfcc_uuid, dfc->dfc_uuid) != 0)
+						continue;
+					dfc->dfs_ino = dfcc->dfcc_ino;
+					break;
+				}
+			}
+		}
+		D_SPIN_UNLOCK(&dfuse_info->di_lock);
 	}
 
-	dfc->dfs_ino = atomic_fetch_add_relaxed(&dfuse_info->di_ino_next, 1);
+	DFUSE_TRA_DEBUG(dfp, "New cont " DF_UUIDF " in pool " DF_UUIDF, DP_UUID(dfc->dfc_uuid),
+			DP_UUID(dfp->dfp_uuid));
+
+	if (dfc->dfs_ino == 0) {
+		dfc->dfs_ino      = atomic_fetch_add_relaxed(&dfuse_info->di_ino_next, 1);
+		dfc->dfc_save_ino = true;
+		DFUSE_TRA_INFO(dfc, "Assigned new inode number %ld", dfc->dfs_ino);
+
+	} else {
+		DFUSE_TRA_INFO(dfc, "Reusing inode number %ld", dfc->dfs_ino);
```

Comment:
do we need to update the historic list by removing the entry from there?

Reply:
No.  In theory we could remove it from the historic list and it would all work, but then when this new container was closed then it would get added to the historic list so I don't see any point adding/removing it.  It would complicate the _ch_free() logic on close to make the decision there as well.